src/masternodeman.cpp

```diff
@@ -644,6 +671,401 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
+        // that was NOT a good idea to cheat and verify itself,
+        // ban the node we received such message from
+        Misbehaving(pnode->id, 100);
+        return;
+    }
+
+    uint256 blockHash;
+    if(!GetBlockHash(blockHash, mnv.nBlockHeight)) {
+        // this shouldn't happen...
+        LogPrintf("MasternodeMan::ProcessVerifyBroadcast -- Can't get block hash for unknown block height %d, peer=%d\n", mnv.nBlockHeight, pnode->id);
+        return;
+    }
+
+    int nRank = GetMasternodeRank(mnv.vin2, pCurrentBlockIndex->nHeight, MIN_POSE_PROTO_VERSION);
+    if(nRank < MAX_POSE_RANK) {
+        LogPrint("masternode", "MasternodeMan::ProcessVerifyBroadcast -- Mastrernode is not in top %d, current rank %d, peer=%d\n",
+                    MAX_POSE_RANK, nRank, pnode->id);
+        return;
+    }
+
+    {
+        LOCK(cs);
+
+        std::string strMessage1 = strprintf("%s%d%s", mnv.addr.ToString(false), mnv.nonce, blockHash.ToString());
+        std::string strMessage2 = strprintf("%s%d%s%s%s", mnv.addr.ToString(false), mnv.nonce, blockHash.ToString(),
+                                mnv.vin1.prevout.ToStringShort(), mnv.vin2.prevout.ToStringShort());
+
+        CMasternode* pmn1 = Find(mnv.vin1);
+        if(!pmn1) {
+            LogPrintf("CMasternodeMan::ProcessVerifyBroadcast -- can't find masternode1 %s\n", mnv.vin1.prevout.ToStringShort());
+            return;
+        }
+
+        CMasternode* pmn2 = Find(mnv.vin2);
+        if(!pmn2) {
+            LogPrintf("CMasternodeMan::ProcessVerifyBroadcast -- can't find masternode2 %s\n", mnv.vin2.prevout.ToStringShort());
+            return;
+        }
+
+        if(pmn1->addr != mnv.addr) {
+            LogPrintf("CMasternodeMan::ProcessVerifyBroadcast -- addr %s do not match %s\n", mnv.addr.ToString(), pnode->addr.ToString());
+            return;
+        }
+
+        if(darkSendSigner.VerifyMessage(pmn1->pubKeyMasternode, mnv.vchSig1, strMessage1, strError)) {
+            LogPrintf("MasternodeMan::ProcessVerifyBroadcast -- VerifyMessage() for masternode1 failed, error: %s\n", strError);
+            return;
+        }
+
+        if(darkSendSigner.VerifyMessage(pmn2->pubKeyMasternode, mnv.vchSig2, strMessage2, strError)) {
+            LogPrintf("MasternodeMan::ProcessVerifyBroadcast -- VerifyMessage() for masternode2 failed, error: %s\n", strError);
+            return;
+        }
+
+        if(!pmn1->IsPoSeVerified()) {
+            pmn1->nPoSeBanScore--;
+        }
+        mnv.Relay();
+
+        LogPrintf("CMasternodeMan::ProcessVerifyBroadcast -- verified masternode %s for addr %s\n",
+                    pmn1->vin.prevout.ToStringShort(), pnode->addr.ToString());
+
+        // increase ban score for everyone else
+        int nCount = 0;
+        BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+            if(mn.addr != mnv.addr || mn.vin.prevout == mnv.vin1.prevout) continue;
+            mn.nPoSeBanScore++;
```

Comment:
Why are you increasing other nodes PoS scores here ?  Since the max is only 5 isn't this going to cause a lot of MN's to be banned ?